charms/ambassador-auth/reactive/ambassador_auth.py

```diff
@@ -0,0 +1,94 @@
+import hashlib
+import os
+import random
+import string
+from base64 import b64encode
+
+import bcrypt
+import yaml
+from charmhelpers.core import hookenv
+from charms.reactive import set_flag, clear_flag, when, when_not
+
+from charms import layer
+
+
+@when('charm.ambassador-auth.started')
+def charm_ready():
+    layer.status.active('')
+
+
+@when('layer.docker-resource.oci-image.changed', 'config.changed')
+def update_image():
+    clear_flag('charm.ambassador-auth.started')
+
+
+@when('layer.docker-resource.oci-image.available')
+@when_not('charm.ambassador-auth.started')
+def start_charm():
+    layer.status.maintenance('configuring container')
+
+    image_info = layer.docker_resource.get_info('oci-image')
+    service_name = hookenv.service_name()
+
+    port = hookenv.config('port')
+
+    username = hookenv.config('username') or 'admin'
+    password = hookenv.config('password')
+
+    if not password:
+        password = ''.join(random.choices(string.ascii_uppercase + string.digits, k=32))
+
+    hookenv.log(f'PASSWORD: {password}')
```

Comment:
@kwmonroe: Any better ideas for how to expose a randomly-generated password to the user? I asked about this over in here, but haven't gotten a response yet:

https://discourse.jujucharms.com/t/how-do-i-handle-passwords-in-kubernetes-charms/1271

Reply:
@knkski unfortunately, no. Some of this goes back to an implicit trust you have with model operators. No matter the safeguards done by the charms, if you need expose something secret like this, model ops can discover it.

Is your specific concern here in the way you're calling hookenv.log, or more generically about "how to get a secret to an operator"?

Reply:
@kwmonroe: Yes. Logs may get sent places where people that shouldn't be able to see the password will be able to see it, and also it's not great UI to have the official method be `juju debug-log --include ambassador-auth/0 | grep PASSWORD`, particularly because that unit number can change.

Reply:
@knkski some non-k8s charms (grafana, graylog) write the random password to a file and then have an action or `juju run` invocation for users to fetch that file.  That won't work for k8s charms, so i think you may be stuck making the password a required config option.